Add C-c and 2x C-c support

stackmystack opened the pull request:
Hello Jeremy,

I am submitting something I'm using internally for a while now and it's really useful to my use case.

Often, hitting C-c vomits pages and pages of stack traces. 

What I want is to allow the test suite to stop as soon as possible, as if I asked for a fail fast and something failed (but not really), so I could get the reports at the end … and If I really want to quit, I just hit C-c a second time, and _I should **NOT** get pages of stack traces_.

I didn't include any tests because I don't know if you'd want to include this feature in the first place, and to be quite frank, I am not so sure how I would go about it.

Anyway, I tried to copy your work on fail fast, and make it opt-in.

stackmystack committed:
Refactor fail_fast in preparation for interrupt support

stackmystack committed:
Add support for graceful halting via interrupts

jeremyevans commented on lib/minitest/parallel_fork/ctrl_c.rb:
```diff
@@ -5,30 +5,35 @@ module Minitest::ParalleForkCtrlC
-            # Process already terminated
-          end
+      if @force_exit
+        child_pids.each do |pid|
```
We should probably DRY up this and the loop below by adding a method that takes the signal, and sends it to each pid.

There also is a naming difference here `child_pids` vs. `children_pids` below.   However, if you adjust for the recommendations below, you can use `@parallel_fork_child_pids`.

jeremyevans commented on lib/minitest/parallel_fork/halt.rb:
```diff
@@ -0,0 +1,80 @@
+          results << parallel_fork_data_from_marshal(thread.value)
+
+          if @parallel_fork_stop
+            # If any child failed fast, signal other children to fail fast
```
You may want to update this comment and all comments in this file in a new commit, since this is no longer just for fast failures.

jeremyevans commented on lib/minitest/parallel_fork/ctrl_c.rb:
```diff
@@ -5,30 +5,35 @@ module Minitest::ParalleForkCtrlC
     Signal.trap(:USR1) do
       @parallel_fork_stop = true
     end
-    Signal.trap(:INT, 'IGNORE')
```
Does the order matter here?

stackmystack replied:
Absolutely not.

jeremyevans commented on lib/minitest/parallel_fork/fail_fast.rb:
```diff
@@ -1,69 +1,15 @@
 module Minitest::ParalleForkFailFast
+  include Minitest::Halt
+
   def run_after_parallel_fork_hook(i)
```
Seems like this should be moved to `halt.rb` as well.  It's used by both `fail_fast.rb` and `ctrl_c.rb`, so it would avoid duplication of code.

jeremyevans commented on lib/minitest/parallel_fork/fail_fast.rb:
```diff
@@ -1,69 +1,15 @@
+require_relative './halt'
```
You can use `halt` instead of `./halt` here and in the other file.

jeremyevans commented on lib/minitest/parallel_fork/ctrl_c.rb:
```diff
@@ -5,30 +5,35 @@ module Minitest::ParalleForkCtrlC
+        @force_exit = true
+        puts "\nInterrupted."
+        puts 'Exiting ...'
+        puts 'C^c again to exit now.'
```
`Interrupt again to exit immediately.`

jeremyevans reviewed (changes requested):
Thank you for your patience.  I had time today to properly review this.  I think some changes need to be made before I can accept this.  Would you be willing to work on them?

stackmystack commented:
Thanks, I will check everything this week.

stackmystack commented:
Thanks for the review.

I addressed all your points, except for factoring out all the forking block, because I didn't manage to do it properly; I added a different method that made it cleaner though.

I tested on my local project, by hand, and this version works. Unit tests are still missing though.

jeremyevans commented:
For testing, I think you could add an example test suite with a bunch of tests that only sleep, and then send SIGINT to the master process and make sure all worker processes clean up.

jeremyevans commented:
This looks good.  Thanks for working on it.  I plan to merge it shortly, with some changes on top.

jeremyevans commented:
Thanks for the reminder. I can probably release later this week.
